File: berryc/syntax/lexer.go
package syntax

import (
	"berryc/report"
	"bufio"
	"io"
	"strings"
	"unicode"
)

type Lexer struct {
	file        *bufio.Reader
	modName     string
	displayPath string

	tokBuff *strings.Builder

	line, col           int
	startLine, startCol int

	ahead rune
}

func NewLexer(file io.Reader, modName, displayPath string) *Lexer {
	return &Lexer{
		file:        bufio.NewReader(file),
		modName:     modName,
		displayPath: displayPath,
		tokBuff:     &strings.Builder{},
		line:        1, col: 1,
		startLine: 1, startCol: 1,
		ahead: 0,
	}
}

func (l *Lexer) NextToken() *Token {
	for l.peek() {
		switch l.ahead {
		case '\t', ' ', '\r', '\n':
			l.skip()
		case '/':
			l.mark()
			l.read()

			if l.peek() {
				if l.ahead == '/' {
					l.skipLineComment()
					continue
				} else if l.ahead == '*' {
					l.skipBlockComment()
					continue
				}
			}

			return l.makeToken(TOK_FSLASH)
		case '\'':
			return l.lexRuneLit()
		case '"':
			return l.lexStringLit()
		case '+':
			return l.lexSingle(TOK_PLUS)
		case '-':
			return l.lexSingle(TOK_MINUS)
		case '*':
			return l.lexSingle(TOK_STAR)
		case '%':
			return l.lexSingle(TOK_MOD)
		case '&':
			return l.lexSingle(TOK_AMP)
		case '|':
			return l.lexSingle(TOK_PIPE)
		case '^':
			return l.lexSingle(TOK_CARRET)
		case '=':
			return l.lexSingle(TOK_ASSIGN)
		case '(':
			return l.lexSingle(TOK_LPAREN)
		case ')':
			return l.lexSingle(TOK_RPAREN)
		case '{':
			return l.lexSingle(TOK_LBRACE)
		case '}':
			return l.lexSingle(TOK_RBRACE)
		case '[':
			return l.lexSingle(TOK_LBRACKET)
		case ']':
			return l.lexSingle(TOK_RBRACKET)
		case '.':
			return l.lexSingle(TOK_DOT)
		case ',':
			return l.lexSingle(TOK_COMMA)
		case ':':
			return l.lexSingle(TOK_COLON)
		case ';':
			return l.lexSingle(TOK_SEMICOLON)
		case '@':
			return l.lexSingle(TOK_ATSIGN)
		default:
			if unicode.IsLetter(l.ahead) || l.ahead == '_' {
				return l.lexIdentOrKeyword()
			} else if '0' <= l.ahead && l.ahead <= '9' {
				return l.lexNumberLit()
			} else {
				l.mark()
				l.error("unknown character")
			}
		}
	}

	return &Token{Kind: TOK_EOF}
}

/* -------------------------------------------------------------------------- */

func (l *Lexer) lexNumberLit() *Token {
	l.mark()

	isFloat := false
	base := 10

	if l.ahead == '0' {
		l.read()

		if l.peek() {
			switch l.ahead {
			case 'b':
				base = 2
				l.read()
			case 'o':
				base = 8
				l.read()
			case 'x':
				base = 16
				l.read()
			}
		}
	}

	switch base {
	case 2:
		for l.peek() {
			if l.ahead == '0' || l.ahead == '1' {
				l.read()
			} else if l.ahead == '_' {
				l.skip()
			} else {
				break
			}
		}
	case 8:
		for l.peek() {
			if '0' <= l.ahead && l.ahead < '9' {
				l.read()
			} else if l.ahead == '_' {
				l.skip()
			} else {
				break
			}
		}
	case 10:
		isFloat = l.readBase10NumLit()
	case 16:
		isFloat = l.readBase16NumLit()
	}

	if isFloat {
		return l.makeToken(TOK_FLOATLIT)
	} else {
		return l.makeToken(TOK_INTLIT)
	}
}

func (l *Lexer) readBase10NumLit() (isFloat bool) {
	hasExp := false
	expectDigit := false

	for l.peek() {
		switch l.ahead {
		case '.':
			if hasExp {
				l.error("exponent in decimal")
			} else if isFloat {
				l.error("multiple decimals in float literal")
			}

			isFloat = true
			l.read()
		case 'e', 'E':
			if hasExp {
				l.error("multiple exponents in decimal")
			}

			isFloat = true
			hasExp = true
			l.read()

			if l.peek() && l.ahead == '-' {
				l.read()

				expectDigit = true
			}
		case '_':
			l.skip()
		default:
			if '0' <= l.ahead && l.ahead <= '9' {
				l.read()
				expectDigit = false
			} else if expectDigit {
				l.error("expected digit")
			} else {
				return
			}
		}
	}

	if expectDigit {
		l.error("expected digit")
	}

	return
}

func (l *Lexer) readBase16NumLit() (isFloat bool) {
	hasExp := false
	expectDigit := false

	for l.peek() {
		switch l.ahead {
		case '.':
			if hasExp {
				l.error("exponent in decimal")
			} else if isFloat {
				l.error("multiple decimals in float literal")
			}

			isFloat = true
			l.read()
		case 'p', 'P':
			if hasExp {
				l.error("multiple exponents in decimal")
			}

			isFloat = true
			hasExp = true
			l.read()

			if l.peek() && l.ahead == '-' {
				l.read()

				expectDigit = true
			}
		case '_':
			l.skip()
		default:
			if '0' <= l.ahead && l.ahead <= '9' || 'a' <= l.ahead && l.ahead < 'g' || 'A' <= l.ahead && l.ahead < 'G' {
				l.read()
				expectDigit = false
			} else if expectDigit {
				l.error("expected digit")
			} else {
				return
			}
		}
	}

	if expectDigit {
		l.error("expected digit")
	}

	return
}

/* -------------------------------------------------------------------------- */

var keywordPatterns = map[string]TokenKind{
	"func": TOK_FUNC,
	"let":  TOK_LET,
	"as":   TOK_AS,

	"i8":   TOK_I8,
	"u8":   TOK_U8,
	"i16":  TOK_I16,
	"u16":  TOK_U16,
	"i32":  TOK_I32,
	"u32":  TOK_U32,
	"i64":  TOK_I64,
	"u64":  TOK_U64,
	"f32":  TOK_F32,
	"f64":  TOK_F64,
	"bool": TOK_BOOL,

	"null":  TOK_NULL,
	"true":  TOK_BOOLLIT,
	"false": TOK_BOOLLIT,
}

func (l *Lexer) lexIdentOrKeyword() *Token {
	l.mark()
	l.read()

	for l.peek() && (unicode.IsLetter(l.ahead) || unicode.IsDigit(l.ahead) || l.ahead == '_') {
		l.read()
	}

	if kkind, ok := keywordPatterns[l.tokBuff.String()]; ok {
		return l.makeToken(kkind)
	}

	return l.makeToken(TOK_IDENT)
}

/* -------------------------------------------------------------------------- */

func (l *Lexer) lexSingle(kind TokenKind) *Token {
	l.mark()
	l.read()
	return l.makeToken(kind)
}

/* -------------------------------------------------------------------------- */

func (l *Lexer) lexStringLit() *Token {
	l.mark()
	l.skip()

	for l.peek() {
		switch l.ahead {
		case '"':
			l.skip()
			return l.makeToken(TOK_STRLIT)
		case '\n':
			l.error("string literal contains new line")
		case '\\':
			l.readEscapeCode()
		default:
			l.read()
		}
	}

	l.error("unclosed string literal")
	return nil
}

func (l *Lexer) lexRuneLit() *Token {
	l.mark()
	l.skip()

	if !l.peek() {
		l.error("unclosed rune literal")
	} else if l.ahead == '\'' {
		l.error("empty rune literal")
	} else if l.ahead == '\\' {
		l.readEscapeCode()
	} else {
		l.read()
	}

	if !l.peek() {
		l.error("unclosed rune literal")
	} else if l.ahead != '\'' {
		l.error("rune literal contains multiple characters")
	} else {
		l.skip()
	}

	return l.makeToken(TOK_RUNELIT)
}

func (l *Lexer) readEscapeCode() {
	l.read()

	if !l.peek() {
		l.error("expected escape code")
	}

	switch l.ahead {
	case 'a', 'b', 'f', 'n', 'r', 't', 'v', '0', '\'', '"':
		l.read()
	default:
		l.error("invalid escape code")
	}
}

/* -------------------------------------------------------------------------- */

func (l *Lexer) skipLineComment() {
	l.tokBuff.Reset()
	l.skip()

	for l.peek() {
		l.skip()

		if l.ahead == '\n' {
			break
		}
	}
}

func (l *Lexer) skipBlockComment() {
	l.tokBuff.Reset()
	l.skip()

	for l.peek() {
		l.skip()

		if l.ahead == '*' {
			if l.peek() && l.ahead == '/' {
				l.skip()
				break
			}
		}
	}
}

/* -------------------------------------------------------------------------- */

func (l *Lexer) makeToken(kind TokenKind) *Token {
	tok := &Token{
		Kind:  kind,
		Value: l.tokBuff.String(),
		Span:  l.getSpan(),
	}

	l.tokBuff.Reset()

	return tok
}

func (l *Lexer) mark() {
	l.startLine = l.line
	l.startCol = l.col
}

func (l *Lexer) error(msg string) {
	report.Throw(&report.SourceError{
		Message: msg,
		Info: &report.SourceInfo{
			ModName:     l.modName,
			DisplayPath: l.displayPath,
			Span:        l.getSpan(),
		},
	})
}

func (l *Lexer) getSpan() *report.TextSpan {
	return &report.TextSpan{
		StartLine: l.startLine, StartCol: l.startCol,
		EndLine: l.line, EndCol: l.col,
	}
}

/* -------------------------------------------------------------------------- */

func (l *Lexer) peek() bool {
	r, _, err := l.file.ReadRune()
	if err != nil {
		if err == io.EOF {
			return false
		}

		report.Throw(err)
	}

	l.ahead = r
	l.file.UnreadRune()

	return true
}

func (l *Lexer) read() bool {
	r, _, err := l.file.ReadRune()
	if err != nil {
		if err == io.EOF {
			return false
		}

		report.Throw(err)
	}

	l.tokBuff.WriteRune(r)
	l.updatePos(r)

	return true
}

func (l *Lexer) skip() bool {
	r, _, err := l.file.ReadRune()
	if err != nil {
		if err == io.EOF {
			return false
		}

		report.Throw(err)
	}

	l.updatePos(r)

	return true
}

func (l *Lexer) updatePos(r rune) {
	switch r {
	case '\n':
		l.line++
		l.col = 1
	case '\t':
		l.col += 4
	default:
		l.col += 1
	}
}

File: src/syntax/lexer.cpp
#include "lexer.hpp"

#include <stdarg.h>
#include <ctype.h>
#include <unordered_map>

Lexer::Lexer(std::ifstream& file_, const SourceFile& src_file_)
: file(file_)
, src_file(src_file_)
, line(1)
, col(1)
, start_line(1)
, start_col(1)
, rlen(0)
{}

void Lexer::NextToken(Token& tok) {
    while (peek()) {
        switch (ahead) {
        case '\n':
        case '\t':
        case '\r':
        case ' ':
            skip();
            break;
        case '/':
            mark();
            read();

            if (peek()) {
                if (ahead == '/') {
                    skipLineComment();
                    continue;
                } else if (ahead == '*') {
                    skipBlockComment();
                    continue;
                } else if (ahead == '=') {
                    read();
                    makeToken(tok, TOK_FSLASH_ASSIGN);
                    return;
                }
            }

            makeToken(tok, TOK_FSLASH);
            return;
        case '\'':
            lexRuneLit(tok);
            return;
        case '"':
            lexStrLit(tok);
            return;
        case '+':
            lexSingleWithAssignOrDouble(tok, TOK_PLUS, TOK_INC, TOK_PLUS_ASSIGN);
            return;
        case '-':
            lexSingleWithAssignOrDouble(tok, TOK_MINUS, TOK_DEC, TOK_MINUS_ASSIGN);
            return;
        case '*':
            lexSingleOrAssign(tok, TOK_STAR, TOK_STAR_ASSIGN);
            return;
        case '%':
            lexSingleOrAssign(tok, TOK_MOD, TOK_MOD_ASSIGN);
            return;
        case '<':
            lexSingleOrDoubleWithAssign(tok, TOK_LT, TOK_SHL, TOK_LE, TOK_SHL_ASSIGN);
            return;
        case '>':
            lexSingleOrDoubleWithAssign(tok, TOK_GT, TOK_SHR, TOK_GE, TOK_SHR_ASSIGN);
            return;
        case '&':
            lexSingleOrDoubleWithAssign(tok, TOK_AMP, TOK_AND, TOK_AMP_ASSIGN, TOK_AND_ASSIGN);
            return;
        case '|':
            lexSingleOrDoubleWithAssign(tok, TOK_PIPE, TOK_OR, TOK_PIPE_ASSIGN, TOK_OR_ASSIGN);
            return;
        case '^':
            lexSingleOrAssign(tok, TOK_CARRET, TOK_CARRET_ASSIGN);
            return;
        case '~':
            lexSingle(tok, TOK_TILDE);
            return;
        case '!':
            lexSingleOrAssign(tok, TOK_NOT, TOK_NE);
            return;
        case '=':
            lexSingleOrDouble(tok, TOK_ASSIGN, TOK_EQ);
            return;
        case '(':
            lexSingle(tok, TOK_LPAREN);
            return;
        case ')':
            lexSingle(tok, TOK_RPAREN);
            return;
        case '[':
            lexSingle(tok, TOK_LBRACKET);
            return;
        case ']':
            lexSingle(tok, TOK_RBRACKET);
            return;
        case '{':
            lexSingle(tok, TOK_LBRACE);
            return;
        case '}':
            lexSingle(tok, TOK_RBRACE);
            return;
        case ',':
            lexSingle(tok, TOK_COMMA);
            return;
        case '.':
            lexSingle(tok, TOK_DOT);
            return;
        case ';':
            lexSingle(tok, TOK_SEMI);
            return;
        case ':':
            lexSingle(tok, TOK_COLON);
            return;
        case '@':
            lexSingle(tok, TOK_ATSIGN);
            return;
        case '#':
            lexDirective(tok);
            return;
        default:
            if (isdigit(ahead)) {
                lexNumberLit(tok);
                return;
            } else if (isalpha(ahead) || ahead == '_') {
                lexKeywordOrIdent(tok);
                return;
            } else {
                fatal("unknown rune: U+{:X}\n", ahead);
            }
        }
    }

    tok.kind = TOK_EOF;
}

/* -------------------------------------------------------------------------- */

static std::unordered_map<std::string, TokenKind> keyword_patterns {
    { "let", TOK_LET },
    { "const", TOK_CONST },
    { "func", TOK_FUNC },
    { "struct", TOK_STRUCT },
    { "enum", TOK_ENUM },
    { "type", TOK_TYPE },
    { "factory", TOK_FACTORY },
    { "if", TOK_IF },
    { "elif", TOK_ELIF },
    { "else", TOK_ELSE },
    { "while", TOK_WHILE },
    { "for", TOK_FOR },
    { "do", TOK_DO },
    { "match", TOK_MATCH },
    { "case", TOK_CASE },
    { "unsafe", TOK_UNSAFE },
    { "break", TOK_BREAK },
    { "continue", TOK_CONTINUE },
    { "return", TOK_RETURN },
    { "fallthrough", TOK_FALLTHROUGH },
    { "new", TOK_NEW },
    { "as", TOK_AS },
    { "null", TOK_NULL },
    { "i8", TOK_I8 },
    { "u8", TOK_U8 },
    { "i16", TOK_I16 },
    { "u16", TOK_U16 },
    { "i32", TOK_I32 },
    { "u32", TOK_U32 },
    { "i64", TOK_I64 },
    { "u64", TOK_U64 },
    { "f32", TOK_F32 },
    { "f64", TOK_F64 },
    { "bool", TOK_BOOL },
    { "unit", TOK_UNIT },
    { "string", TOK_STRING },
    { "module", TOK_MODULE },
    { "import", TOK_IMPORT },
    { "pub", TOK_PUB },
    { "true", TOK_BOOLLIT },
    { "false", TOK_BOOLLIT }
};

void Lexer::lexKeywordOrIdent(Token& tok) {
    mark();
    read();

    while (peek() && (isalnum(ahead) || ahead == '_')) {
        read();
    }

    auto it = keyword_patterns.find(tok_buff);
    if (it != keyword_patterns.end()) {
        makeToken(tok, it->second);
    } else {
        makeToken(tok, TOK_IDENT);
    }
}

void Lexer::lexDirective(Token& tok) {
    mark();
    skip();

    while (peek() && (isalpha(ahead) || ahead == '_')) {
        read();
    }

    if (tok_buff.size() == 0) {
        fatal("expected directive name");
    }

    makeToken(tok, TOK_DIRECTIVE);
}

/* -------------------------------------------------------------------------- */

static bool isHexDigit(rune r) {
    return isdigit(r) || ('A' <= r && r < 'G') || ('a' <= r && r < 'g');
}

static bool isDecDigit(rune r) {
    return isdigit(r);
}

void Lexer::lexNumberLit(Token& tok) {
    mark();
    read();

    int base = 10;
    if (peek()) {
        switch (ahead) {
        case 'b':
            read();
            base = 2;
            break;
        case 'o':
            read();
            base = 8;
            break;
        case 'x':
            read();
            base = 16;
            break;
        default:
            break;
        }
    }

    bool is_float = false;
    switch (base) {
    case 2:
        while (peek()) {
            if (ahead == '0' || ahead == '1') {
                read();
            } else if (ahead == '_') {
                skip();
            } else {
                break;
            }
        }
        break;
    case 8:
        while (peek()) {
            if ('0' <= ahead && ahead < '9') {
                read();
            } else if (ahead == '_') {
                skip();
            } else {
                break;
            }
        }
        break;
    case 10:
        is_float = readFloatOrIntLit(tok, isDecDigit, 'e', false);
        break;
    case 16:
        is_float = readFloatOrIntLit(tok, isHexDigit, 'p', true);
        break;
    }

    if (is_float) {
        makeToken(tok, TOK_FLOATLIT);
    } else {
        makeToken(tok, TOK_INTLIT);
    }
}

bool Lexer::readFloatOrIntLit(Token& tok, DigitCheckFunc f_is_digit, char exp_char_lower, bool expect_digit) {
    bool is_float = false, has_exp = false;
    
    char exp_char_upper = toupper(exp_char_lower);

    while (peek()) {
        if (ahead == '.') {
            if (has_exp) {
                fatal("decimal point cannot occur in exponent");
            } else if (is_float) {
                fatal("multiple decimal points in float literal");
            }

            is_float = true;
            expect_digit = true;
            read();
        } else if (ahead == '_') {
            skip();
            continue;
        } else if (f_is_digit(ahead)) {
            expect_digit = false;
            read();
        } else if (ahead == exp_char_lower || ahead == exp_char_upper) {
            if (has_exp) {
                fatal("multiple exponents in float literal");
            } 

            has_exp = true;
            is_float = true;
            expect_digit = true;
            read();

            if (peek() && ahead == '-') {
                read();
            }
        } else {
            break;
        }
    }

    if (expect_digit) {
        fatal("expected digit to end float literal");
    }

    return is_float;
}

/* -------------------------------------------------------------------------- */

void Lexer::lexSingleOrDoubleWithAssign(
    Token& tok, 
    TokenKind single, TokenKind doub, 
    TokenKind singleAssign, TokenKind doubAssign
) {
    mark();

    rune first = ahead;
    read();

    if (peek()) {
        if (ahead == first) {
            read();

            if (peek() && ahead == '=') {
                read();
                makeToken(tok, doubAssign);
            } else {
                makeToken(tok, doub);
            }

            return;
        } else if (ahead == '=') {
            read();
            makeToken(tok, singleAssign);
            return;
        }
    }

    makeToken(tok, single);
}

void Lexer::lexSingleWithAssignOrDouble(Token& tok, TokenKind single, TokenKind doub, TokenKind assign) {
    mark();

    rune first = ahead;
    read();

    if (peek()) {
        if (ahead == first) {
            read();
            makeToken(tok, doub);
            return;
        } else if (ahead == '=') {
            read();
            makeToken(tok, assign);
            return;
        }
    }

    makeToken(tok, single);
}

void Lexer::lexSingleOrAssign(Token& tok, TokenKind single, TokenKind assign) {
    mark();
    read();

    if (peek() && ahead == '=') {
        read();
        makeToken(tok, assign);
        return;
    }

    makeToken(tok, single);
}

void Lexer::lexSingleOrDouble(Token& tok, TokenKind single, TokenKind doub) {
    mark();

    rune first = ahead;
    read();

    if (peek() && ahead == first) {
        read();
        makeToken(tok, doub);
        return;
    }

    makeToken(tok, single);
}

void Lexer::lexSingle(Token& tok, TokenKind kind) {
    mark();
    read();
    makeToken(tok, kind);
}

/* -------------------------------------------------------------------------- */

void Lexer::lexStrLit(Token& tok) {
    mark();
    skip();

    while (peek()) {
        if (ahead == '\n') {
            break;
        } else if (ahead == '\"') {
            skip();
            makeToken(tok, TOK_STRLIT);
            return;
        } else if (ahead == '\\') {
            readEscapeSeq();
        } else {
            read();
        }
    }

    fatal("unclosed string literal");
}

void Lexer::lexRuneLit(Token& tok) {
    mark();
    skip();

    if (!peek()) {
        fatal("unclosed rune literal");
    } else if (ahead == '\'') {
        fatal("empty rune literal");
    } else if (ahead == '\n') {
        fatal("rune literal can't contain newline");
    } else if (ahead == '\\') {
        readEscapeSeq();
    } else {
        read();
    }

    if (!peek()) {
        fatal("unclosed rune literal");
    } else if (ahead != '\'') {
        fatal("rune literal is too long");
    }

    skip();

    makeToken(tok, TOK_RUNELIT);
}

void Lexer::readEscapeSeq() {
    read();

    if (!peek()) {
        fatal("expected escape sequence");
    }

    switch (ahead) {
    case 'a':
    case 'b':
    case 'f':
    case 'n':
    case 'r':
    case 't':
    case 'v':
    case '0':
    case '\'':
    case '"':
    case '\\':
        read();
        break;
    default:
        skip();
        fatal("invalid escape sequence");
    }
}

/* -------------------------------------------------------------------------- */

void Lexer::skipLineComment() {
    tok_buff.clear();
    skip();

    while (peek() && ahead != '\n') {
        skip();
    }
}

void Lexer::skipBlockComment() {
    tok_buff.clear();
    skip();

    while (peek()) {
        skip();

        if (ahead == '*' && peek() && ahead == '/') {
            skip();
            return;
        }
    }
}

/* -------------------------------------------------------------------------- */

void Lexer::mark() {
    start_line = line;
    start_col = col;
}

void Lexer::makeToken(Token& tok, TokenKind kind) {
    tok.kind = kind;
    tok.value = std::move(tok_buff);
    tok.span = getSpan();
}

/* -------------------------------------------------------------------------- */

rune Lexer::read() {
    rune r;
    if (rlen > 0) {
        r = ahead;
    } else {
        r = getRune();
        if (r == -1) {
            return -1;
        }
    }

    writeRune();
    updatePos(r);
    rlen = 0;

    return r;
}

rune Lexer::skip() {
    rune r;
    if (rlen > 0) {
        r = ahead;
    } else {
        r = getRune();
        if (r == -1) {
            return -1;
        }
    }

    updatePos(r);
    rlen = 0;

    return r;
}

bool Lexer::peek() {
    if (rlen > 0) {
        return true;
    } else {
        ahead = getRune();
        return ahead != -1;
    }
}

void Lexer::updatePos(rune r) {
    switch (r) {
    case '\n':
        line++;
        col = 1;
        break;
    case '\t':
        col += 4;
        break;
    default:
        col++;
        break;
    }
}

/* -------------------------------------------------------------------------- */

rune Lexer::getRune() {
    byte b1 = file.get();
    if (b1 == 0xff) {
        return -1;
    }
    
    rbuff[0] = b1;

    int n_bytes = 0;
    int32_t r;
    if ((b1 & 0x80) == 0) { // 0xxxxxxx
        rlen = 1;
        return b1;
    } else if ((b1 & 0xe0) == 0xc0) { // 110xxxxx
        n_bytes = 1;
        r = b1 & 0x1f;
    } else if ((b1 & 0xf0) == 0xe0) { // 1110xxxx
        n_bytes = 2;
        r = b1 & 0x0f;
    } else if ((b1 & 0xf8) == 0xf0) { // 11110xxx
        n_bytes = 3;
        r = b1 & 0x07;
    } else {
        fatal("malformed rune: invalid leading byte: {}", b1);
    }

    for (int i = 0; i < n_bytes; i++) {
        byte b = file.get();
        if (b == -1) {
            fatal("malformed rune: expected {} bytes; got EOF at {} bytes", n_bytes + 1, i + 1);
        }

        rbuff[i+1] = b;

        r <<= 6;
        r = r | (b & 0x3f);
    }

    rlen = n_bytes + 1;

    return r;
}

void Lexer::writeRune() {
    tok_buff.append(rbuff, rlen);
}

/* -------------------------------------------------------------------------- */

TextSpan Lexer::getSpan() {
    return {start_line, start_col, line, col};
}
